### PR TITLE
Benji/ios 563 create inputhandlercoordinator base

### PR DESCRIPTION
**Pull request by Benjamindodgson**
Created a base class Coordinator to handle all the shared logic around input. Now both the ThreadCoordinator and ConversationListCoordinator conform to it.

### CODE REVIEW

File: iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+//  Created by Benji Dodgson on 3/23/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+import PhotosUI
+import Photos
+import Localization
+
+protocol SwipeableInputControllerHandler where Self: ViewController {
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift:14]: Here's the protocol and Coordinator

File: iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    
+    lazy var captureVC: UIImagePickerController = {
+        let vc = UIImagePickerController()
+        vc.sourceType = .camera
+        vc.allowsEditing = true
+        vc.delegate = self
+        return vc
+    }()
+    
+    let inputHandlerViewController: InputHandlerViewContoller
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift:45]: While not lazy, its needed right away and only created once.

File: iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    
+    func presentAttachements() {
+        let coordinator = AttachmentsCoordinator(router: self.router, deepLink: self.deepLink)
+        
+        self.present(coordinator) { [unowned self] result in
+            self.handle(attachmentOption: result)
+        }
+    }
+    
+    func present<ChildResult>(_ coordinator: PresentableCoordinator<ChildResult>,
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Input/InputHandlerCoordinator.swift:75]: Both share the need for this presentation logic. Probably fixed a lot of bugs by adding this.

File: iOS/Flows/Threads/ThreadCoordinator.swift [hunk 4]
```diff
@@ -9,26 +9,23 @@
...
-
-    lazy var threadVC = ThreadViewController(channelID: self.channelId,
-                                             messageID: self.messageId,
-                                             startingReplyId: self.startingReplyId)
-    let messageId: MessageId
-    let channelId: ChannelId
-    let startingReplyId: MessageId?
+class ThreadCoordinator: InputHandlerCoordinator<ConversationId> {
+    
+    var threadVC: ThreadViewController {
```
Comment (Benjamindodgson) [inline iOS/Flows/Threads/ThreadCoordinator.swift:14]: Im fine with a force unwrap here and a getter.

File: iOS/Flows/Threads/ThreadCoordinator.swift [hunk 5]
```diff
@@ -39,23 +36,14 @@ class ThreadCoordinator: PresentableCoordinator<Void> {
...
-        vc.dismissHandlers.append { [unowned self] in
-            self.threadVC.becomeFirstResponder()
-        }
-        
-        vc.onDidComplete = { _ in
-            vc.dismiss(animated: true, completion: nil)
+    func presentProfile(for person: PersonType) {
+        let coordinator = ProfileCoordinator(with: person, router: self.router, deepLink: self.deepLink)
+        self.present(coordinator) { [unowned self] result in
+            self.finishFlow(with: result)
```
Comment (Benjamindodgson) [inline iOS/Flows/Threads/ThreadCoordinator.swift:46]: Now we close the thread if a conversation was selected, and the list will scroll to the selected cell.

**Review by MartinJibber — APPROVED** [review]
Awesome, great refactor.

**Pull request merged**